### PR TITLE
SAML

### PR DESCRIPTION
**Pull request by SammyPoot**
Includes the field removal changes.

Restructures the code heavily to use SAML logins.

### CODE REVIEW

**Review by kierenbeckett — DISMISSED** [review]
LGTM, left some minor comments that would be nice to address but not essential, biggest thing isn't actually in the PR (validate cert when making POST request).

Some unit tests of the bits of business logic e.g. the regex matching might be nice?

File: app/employee_view_functions.py [hunk 1]
```diff
@@ -49,8 +49,6 @@ def process_employee_information(employee_information):
                 employee_information[emp_info] = 'No'
             elif emp_info == 'workRestrictions':
                 employee_information[emp_info] = 'None'
-            elif emp_info == 'anyLanguagesSpoken':
```
[the remark this answers is not in the document]
Comment (SammyPoot) [reply app/employee_view_functions.py:52]: That (and other changes) relate to this part of the description: 'Includes the field removal changes.'.

File: app/employee_view_router.py [hunk 2]
```diff
@@ -1,31 +1,32 @@
 from app.employee_view_functions import process_employee_information, get_current_job_role
 from app.views import rmt_view, hq_fo_ccs_view, logistics_view, hr_view, fsss_view, recruitment_view
 
+from . import saml, role_matchers
```
[the remark this answers is not in the document]
Comment (SammyPoot) [reply app/employee_view_router.py]: Fixed

File: app/fieldmapping.py [hunk 3]
```diff
@@ -147,36 +120,5 @@ def map_employee_name(employee_table):
     return employee_name
 
 
-def map_emergency_contact_name(employee_table, both_required=True):
```
[the remark this answers is not in the document]
Comment (SammyPoot) [reply app/fieldmapping.py:150]: Ditto, related to field removal.

File: app/employee_view_router.py [hunk 4]
```diff
@@ -1,31 +1,32 @@
...
+    # Logistics
+    elif role_matchers.logi_regex.match(role_id):
+        return logistics_view.get_employee_tabs(employee_information, cur_job_role, device_information)
+    # HR
+    elif role_matchers.hr_regex.match(role_id):
+        return hr_view.get_employee_tabs(employee_information, cur_job_role, device_information)
+    # FSSS
+    elif role_matchers.fsss_regex.match(role_id):
+        return fsss_view.get_employee_tabs(employee_information, cur_job_role, device_information)
+    # Recruitiment
```
[the remark this answers is not in the document]
Comment (SammyPoot) [reply app/employee_view_router.py]: Fixed.

File: app/credentialhandler.py [hunk 5]
```diff
@@ -34,40 +35,39 @@ def log_entry(request, endpoint):
...
-@credential_routes.view('/')
+#@credential_routes.view('/')
 class Redirect:
     async def get(self, request):
-        raise HTTPFound(request.app.router['Login:get'].url_for())
+        saml.rediect_to_login(request)
 
 
-@credential_routes.view('/signin')
+#@credential_routes.view('/signin')
```
[the remark this answers is not in the document]
Comment (SammyPoot) [reply app/credentialhandler.py]: Fixed

File: app/credentialhandler.py [hunk 6]
```diff
@@ -34,40 +35,39 @@ def log_entry(request, endpoint):
...
     credentials = {'password': password, 'username': user}
-    return requests.post(FSDR_URL + f'/userAuth/checkCredentials', data=credentials,
+    return requests.post(FSDR_URL + f'/userAuth/checkCredentials',
+                         data=credentials,
                          verify=False,
                          auth=HTTPBasicAuth(FSDR_USER, FSDR_PASS))
 
 
-@credential_routes.view('/')
+#@credential_routes.view('/')
```
[the remark this answers is not in the document]
Comment (SammyPoot) [reply app/credentialhandler.py]: Yes, it's been removed.

File: app/credentialhandler.py [hunk 7]
```diff
@@ -34,40 +35,39 @@ def log_entry(request, endpoint):
...
     await clear_stored_search_criteria(session)
 
-def get_fsdr_signin( user, password):
+
+def get_fsdr_signin(user, password):
     credentials = {'password': password, 'username': user}
-    return requests.post(FSDR_URL + f'/userAuth/checkCredentials', data=credentials,
+    return requests.post(FSDR_URL + f'/userAuth/checkCredentials',
+                         data=credentials,
                          verify=False,
```
[the remark this answers is not in the document]
Comment (SammyPoot) [reply app/credentialhandler.py]: That's outside the scope of this PR, please raise a ticket and I will happily deal with it.

File: Dockerfile [hunk 8]
```diff
@@ -1,13 +1,29 @@
...
-RUN pip3 install pipenv && pipenv install --deploy --system
+# user setup
 RUN groupadd -g 984 fsdrui && \
     useradd -r -u 984 -g fsdrui fsdrui
+
+# updates and packages
+RUN apt-get update \
+    && apt-get -y --no-install-recommends install libxml2-dev libxmlsec1-dev libxmlsec1-openssl \
+    && apt-get clean
+RUN pip3 install Flask-Cors requests pipenv
```
[the remark this answers is not in the document]
Comment (SammyPoot) [reply Dockerfile]: Flask-Cors might be possible to delete fully, requests was already in the pipfile. Line cleaned up!

**Review by kierenbeckett — APPROVED** [review]
:+1:

**Pull request merged**